File: 백준/Gold/4195. 친구 네트워크/친구 네트워크.py
import sys
input=sys.stdin.readline

t=int(input())

def findParent(x):
    if parent[x] != x:
        parent[x] = findParent(parent[x])
    return parent[x]

for i in range(t):
    f=int(input()) # 100000이하
    cnt={}
    parent={}
    for j in range(f):
        fr1, fr2 = input().split()
        if fr1 not in parent:
            parent[fr1]=fr1
            cnt[fr1]=1
        if fr2 not in parent:
            parent[fr2]=fr2
            cnt[fr2]=1
        a=findParent(fr1)
        b=findParent(fr2)
        if a!=b:
            parent[a] = b
            cnt[b] += cnt[a]
        print(cnt[b])

File: 백준/Gold/4195. 친구 네트워크/친구 네트워크.java
import java.util.*;
import java.io.*;
public class Main {
    static int TC, F;
    static HashMap<String, String> parents;
    static HashMap<String, Integer> cnt;
    public static String find(String u) {
        if (parents.get(u) == u) return u;
        parents.put(u, find(parents.get(u)));
        return parents.get(u);
    }

    public static void union(String u, String v) {
        parents.put(u, v);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        TC = Integer.parseInt(br.readLine());

        for (int t = 0; t < TC; t++) {
            F = Integer.parseInt(br.readLine());

            parents = new HashMap<>();
            cnt = new HashMap<>();

            for (int i = 0; i < F; i++) {
                StringTokenizer st = new StringTokenizer(br.readLine());
                String fr1 = st.nextToken();
                String fr2 = st.nextToken();

                if (!parents.containsKey(fr1)) {
                    parents.put(fr1, fr1);
                    cnt.put(fr1, 1);
                }

                if (!parents.containsKey(fr2)) {
                    parents.put(fr2, fr2);
                    cnt.put(fr2, 1);
                }

                String a = find(fr1);
                String b = find(fr2);

                if (a != b) {
                    union(a, b);
                    cnt.put(b, cnt.get(a) + cnt.get(b));
                }

                System.out.println(cnt.get(b));
            }

        }
    }
}